readme.md
# DotMatrix5x7
[![License: LGPL v3](https://img.shields.io/badge/License-LGPLv3-blue.svg)](https://www.gnu.org/licenses/lgpl-3.0)
[![Installation instructions](https://www.ardu-badge.com/badge/DotMatrix5x7.svg?)](https://www.ardu-badge.com/DotMatrix5x7)
[![Commits since latest](https://img.shields.io/github/commits-since/felias-fogg/DotMatrix5x7/latest?include_prereleases)](https://github.com/felias-fogg/DotMatrix5x7/commits/master)
![Hit Counter](https://visitor-badge.laobi.icu/badge?page_id=felias-fogg_DotMatrix5x7)
[![Build Status](https://github.com/felias-fogg/DotMatrix5x7/workflows/LibraryBuild/badge.svg)](https://github.com/felias-fogg/DotMatrix5x7/actions)

Drive a single 5x7 dot matrix display directly by Arduino pins (12 needed!)

Version: 1.5.2

Date:  31.05.2023

## Features

* Supports 5x7 dot matrix displays such as TA07-11SEKWA
* Assumes only Arduino pins, no external controller
* Pins and polarity are user definable
* Interrupt driven (using timer 0, optionally timer 1)
* Can scroll vertically and horizontally
* One can use regular strings and strings in flash memory
* It is possible to let characters blink
* You can provide your own delay function (in order to save power or to do something in parallel)

## Usage

This library can be used to control one 5x7 dot matrix display. 12 Arduino pins are necessary to control it (5 column pins, 7 row pins). The rows are multiplexed, which means one needs just 5 resistors for the columns. The display of characters is done in an interrupt driven way. This means you can do other things after you have requested to display a character.

Usually, TIMER0 (the timer for millis and delays) is used for generating the interrupts. Note that this does not interfere with `delay()` or `millis()` since we use a different interrupt! However, it could lead to an interrupt conflict if other parts of the program use the `TIMER0_COMPA` interrupt. For this reason, there exists the compile time option `USETIMER0`, which when defined to be 0 in DotMatrix5x7.h disables the TIMER0 interrupt.

Instead or additionally, the `TimerOne` library by Paul Stoffregen could be used, provided the MCU type is supported by this library.  If you want to use the TimerOne library, `USETIMER1` needs to be 1 (it is 0 by default). Note that TimerOne does not support all MCUs and will generate compilation errors if the MCU is not supported.

At least one of those timers must be usable for the library to function. If both timers are available, one can select which one to use when calling the `begin` method. 

When you want to display a single character `c`, call `Dot5x7.show(c)`. A string can be displayed, e.g., as follows: `Dot5x7.showString("Hello World!", 500, 100)`. This displays the string "Hello World!", where each character is displayed 500 msec and there is a pause of 100 msec between characters. One can also use the `F()` notation to pass a string that is stored in flash memory, i.e., `Dot5x7.showString(F("Hello World!"), 500, 100)`. In addition, one might pass a string stored in flash (which has been defined with PROGMEM) using the `_P`-variant of the method, i.e.: `Dot5x7.showString_P(str, 500, 100)`, provided `str` has been defined as `const char PROGMEM str[] = "Hello World!";`.

In addition, there are methods that scroll a string through the display. Scrolling can be done vertically up or down or horizontally left or right. For example, one can scroll the Hello World string by calling `Dot5x7.scrollLeftString("Hello World!", 300, 50, 1)`. This will scroll the string to the left, displaying each character for 300 msec in the middle of the display and using 50 msec for each shifting step. The final argument of 1 means that one blank column is used between two consecutive characters. There are also methods for scrolling right, up, or down. Further, one can, of course, use the `F()` notation and there are also the `_P` variants of the methods. 

## Configuration

The row and column pins can be specified when calling the `begin` method. In addition, you can specify the polarity of when the row and column pins are active, respectively. Default is that row pins are active LOW and column pins are active HIGH. This corresponds to driving a column anode display directly from the pins. Finally, as a last (default) argument, one can specify the timer (0 or 1). As mentioned, this is only significant if both timers are available for interrupts, i.e., both the compile time constants `USETIMER1` and `USETIMER0` have to be set to 1 in DotMatrix5x7.h.

In addition, there are a few methods that control the interface:

- `setUpsideDown(bool enable)` controls whether the characters are displayed upside-down.
- `setFont(const byte *f)` can be used to select an alternative font.
- `setFramesPerSecond(int fps)` is used to control how often a complete matrix is displayed. The default value is 50 and one should not select values below 42. This is only effective if Timer1 is used.
- `setBlinkFrames(int blinkon, int blinkoff)` controls blinking by specifying the number of frames where the character is displayed and the number of frames when the display is off. If either value is 0, no blinking will happening.
- `setDelayFunction(void (*f) (long unsigned int))` enables one to specify a custom delay function that is to be used when the `showString` and `scrollXXXString` methods are used. This may be used in order to minimize power consumption or to communicate with peripherals.
- `sleep()` stops timer and switches off all LEDs.
- `wakeup()` restarts timer.



## Adding new symbols

If you would like to create your own symbols, you can use the website http://dotmatrixtool.com/. Select the following parameters:

- Width: 6
- Height: 8
- Byte Order: Column Major
- Endian: Little Endian

Now you can can left click (to mark) and right click (to  unmark) and produce your symbol in the left five columns and the upper  seven rows. Press "Generate", and you get the code you have to fill into the font table (the first five bytes). 

Now, you can either use one of the entries in the standard font table in `dotfont5x7.h` or you can setup your own private font  table, which needs to be a PROGMEM char array similar to the `font5x7`  array, and then use the setFont method in order to switch between the  standard table and your new table. Let us assume, you have defined a Yen symbol, which leads to the five-byte code `0x29, 0x2a, 0x7c, 0x2a, 0x29`. Then you could, e.g., declare and use this symbol as follows:

```c++
const unsigned char  money[] PROGMEM = {0x29, 0x2a, 0x7c, 0x2a, 0x29};
...
Dot5x7.setFont(money);  // select custom font
Dot5x7.show(0);         // show the entry in custom font at location 0
delay(1000);
Dot5x7.setFont();       // switch back to system font
...
```



## Example

```c++
// Simple sketch for displaying a character and a string
#include <DotMatrix5x7.h>

void setup()
{
    Dot5x7.begin(0, 1, 2, 3, 4,         // column pins
	         5, 6, 7, 8, 9, 10, 11,     // row pins
	         LOW,                       // value when row pin is active (default value)
		 	 HIGH);                     // value when column pin is active (default value)
    Dot5x7.setUpsideDown(true);         // display all chars upside down
    Dot5x7.setFramesPerSecond(50);      // display 50 frames per second (default value)			 
}	 

void loop()
{
	Dot5x7.show('X'); // display 'X'
	delay(1000);      // for 1 second
	Dot5x7.clear();   // clear display
	delay(1000);      // show empty display for 1 second
	Dot5x7.showString("Hello", 700, 100); // display string, with 0.7sec time for showing each char
	                                      // and 0.1sec pause between two chars.
}
```

This software is published under the [LGPL](http://www.gnu.org/licenses/lgpl-3.0.html)


changelog.txt
Version 1.5.2 (31.5.2023)
Changed sleep and wakeup methods so that they now switch all pins to
input instead of pulling them to ground, which led to excessive
current consumption on a 1634.

Version 1.5.1 (30.5.2023)
Fixed clear method, which did something unexpected when a non-standard
font was activated.

Version 1.5.0 (29.5.2023)
TIMER0 is now the default interrupt source. Added a default parameter to the
begin method for selecting the appropriate timer interrupt. In order
to avoid TIMER0_CMPA interrupt conflicts, one needs to set the compile
time constant USETIMER0 in DotMatrix5x7.h to 0. If one wants to use
TIMER1, one needs to define USETIMER1 to 1. Only if both compile time
constants are set to 1, then the timer parameter for the begin method is significant.

Version 1.4.4
Corrected code for sZET in font file (it is now truly 0xE1)
Added comment for adding new symbols
Added GitHub actions and badges
Made example idleFunction compatible with offical Low-Power lib
Added code in DotMatrix5x7.cpp to use other timing constants and ISRs
when USETIMER0 is defined

Version  1.4.3
New methods: sleep() and wakeup(), which shut down the display, respectively, wakes it up again
New method: setFont() without arguments restores original font
Scrolling is now flicker free
Bug fix: In the initialization, rows and columns were set to HIGH and LOW instead of _rowactive and _columnactive

Version 1.4.2
Typos corrected and font file updated (now all chars correspond to the IBM-PC extended char set)

Version 1.4.1
font file moved to the right place

Version 1.4.0
First published version


src/DotMatrix5x7.h

#ifndef DOTMATRIX5X7_H
#define DOTMATRIX5X7_H

#define USETIMER0 1
#define USETIMER1 0

#include <Arduino.h>
#if USETIMER1
#include <TimerOne.h>
#endif

#if !USETIMER0 && !defined(TIMER1_A_PIN)
#error "Either Timer 0 or Timer 1 needs to be available for generating interrupts for DotMatrix5x7"
#endif

#define NUMCOLS 5
#define NUMROWS 7
#define UPDIR 0
#define DOWNDIR 1
#define RIGHTDIR 2
#define LEFTDIR 3


// Codes for German Umlauts:
#define aUML "\x84" // ä = 0x84
#define oUML "\x94" // ö = 0x94
#define uUML "\x81" // ü = 0x81
#define sZET "\xE0" // ß = 0xE1
#define AUML "\x8E" // Ä = 0x8E
#define OUML "\x99" // Ö = 0x99
#define UUML "\x9A" // Ü = 0x9A

#define DEGREE "\xF8" 

class DotMatrix5x7 {
 public:
  void begin(byte col1, byte col2, byte col3, byte col4, byte col5, byte row1, byte row2, byte row3,
	     byte row4, byte row5, byte row6, byte row7, byte timer=0);
  void begin(byte col1, byte col2, byte col3, byte col4, byte col5, byte row1, byte row2, byte row3,
	     byte row4, byte row5, byte row6, byte row7, byte rowactive, byte columnactive, byte timer=0);
  void setFont(const byte *f);
  void setFont(void);
  byte getTimer(void);
  void setFramesPerSecond(int fps);
  void setUpsideDown(bool enable);
  void setDelayFunction(void (*f) (long unsigned int));
  void setBlinkFrames(int blinkon, int blinkoff);
  void sleep();
  void wakeup();
  void clear(void);
  void show(byte c);
  void showString(const char *str, int ontime, int offtime);
  void showString(const __FlashStringHelper *str, int ontime, int offtime);
  void showString_P(const char *str, int ontime, int offtime);
  void scrollUp(byte curr, byte next, byte offset, byte step);
  void scrollUpString(char const *str, int showtime, int scrolltime,  int offset);
  void scrollUpString(const __FlashStringHelper *str,  int showtime, int scrolltime, int offset);
  void scrollUpString_P(char const *str,  int showtime, int scrolltime, int offset);
  void scrollDown(byte curr, byte next, byte offset, byte step);
  void scrollDownString(char const *str, int showtime, int scrolltime,  int offset);
  void scrollDownString(const __FlashStringHelper *str,  int showtime, int scrolltime, int offset);
  void scrollDownString_P(char const *str,  int showtime, int scrolltime, int offset);
  void scrollLeft(byte curr, byte next, byte offset, byte step);
  void scrollRight(byte curr, byte next, byte offset, byte step);
  void scrollRightString(char const *str,  int showtime, int scrolltime, int offset);
  void scrollRightString(const __FlashStringHelper *str,  int showtime, int scrolltime, int offset);
  void scrollRightString_P(char const *str,  int showtime, int scrolltime, int offset);
  void scrollLeftString(char const *str,  int showtime, int scrolltime, int offset);
  void scrollLeftString(const __FlashStringHelper *str,  int showtime, int scrolltime, int offset);
  void scrollLeftString_P(char const *str,  int showtime, int scrolltime, int offset);
  void scrollString(char const *str,  int showtime, int scrolltime, int offset, byte dir, bool flash=false);
  static void displayRow(void);

 private:
  void initMatrix(void);
  void showStringWorker(char const *str, int ontime, int offtime, bool flash);
  void scrollDisplayDown(byte curr, byte next, byte offset, byte step);
  void scrollDisplayUp(byte curr, byte next, byte offset, byte step);
  void scrollDisplayRight(byte curr, byte next, byte offset, byte step);
  void scrollDisplayLeft(byte curr, byte next, byte offset, byte step);
  void transposePattern(byte cols[], volatile byte rows[]);
  bool _upsidedown = false;
  unsigned long _period;
  byte _rowpin[NUMROWS];
  byte _colpin[NUMCOLS];
  byte *_font;
  byte _rowactive = LOW;
  byte _columnactive = HIGH;
  byte _timer = 0;
  void (*_delay) (long unsigned int) = delay;
  volatile byte _row[NUMROWS];
  volatile byte _currrow = 0;
  volatile bool _blocked = false;
  volatile byte _blank = 0;
  volatile int _blinkon = 0;
  volatile int _blinkoff = 0;
  volatile int _blinkcnt = 0;
};

extern DotMatrix5x7 Dot5x7;
#endif


src/dotfont5x7.h
#include <avr/io.h>
#include <avr/pgmspace.h> 
 
#ifndef DOTFONT5X7_H
#define DOTFONT5X7_H


// standard ascii 5x7 font incl. the IPM-PC extension:
// already defined in DotMatrx5x7.h
//#define aUML "\x84" // ä = 0x84
//#define oUML "\x94" // ö = 0x94
//#define uUML "\x81" // ü = 0x81
//#define sZET "\xE1" // ß = 0xE1
//#define AUML "\x8E" // Ä = 0x8E
//#define OUML "\x99" // Ö = 0x99
//#define UUML "\x9A" // Ü = 0x9A

const unsigned char  font5x7[] PROGMEM = {
  0x00, 0x00, 0x00, 0x00, 0x00, // 0x00 empty
  0x3E, 0x5B, 0x4F, 0x5B, 0x3E, // 0x01 sad
  0x3E, 0x6B, 0x4F, 0x6B, 0x3E, // 0x02 happy
  0x1C, 0x3E, 0x7C, 0x3E, 0x1C, // 0x03 heart
  0x18, 0x3C, 0x7E, 0x3C, 0x18, // 0x04 diamond
  0x1C, 0x57, 0x7D, 0x57, 0x1C, // 0x05 clover
  0x1C, 0x5E, 0x7F, 0x5E, 0x1C, // 0x06 spade
  0x00, 0x18, 0x3C, 0x18, 0x00, // 0x07 small full circle
  0xFF, 0xE7, 0xC3, 0xE7, 0xFF, // 0x08 inverted small circle 
  0x00, 0x18, 0x24, 0x18, 0x00, // 0x09 small empty circle 
  0xFF, 0xE7, 0xDB, 0xE7, 0xFF, // 0x0A block (linefeed)
  0x30, 0x48, 0x3A, 0x06, 0x0E, // 0x0B male
  0x26, 0x29, 0x79, 0x29, 0x26, // 0x0C female
  0x40, 0x7F, 0x05, 0x05, 0x07, // 0x0D note (carriage return)
  0x40, 0x7F, 0x05, 0x25, 0x3F, // 0x0E two notes
  0x5A, 0x3C, 0xE7, 0x3C, 0x5A, // 0x0F sun
  0x7F, 0x3E, 0x1C, 0x1C, 0x08, // 0x10 right triangle
  0x08, 0x1C, 0x1C, 0x3E, 0x7F, // 0x11 left triangle
  0x14, 0x22, 0x7F, 0x22, 0x14, // 0x12 up/down arrow
  0x5F, 0x5F, 0x00, 0x5F, 0x5F, // 0x13 double exclamation mark
  0x06, 0x09, 0x7F, 0x01, 0x7F, // 0x14 paragraph end sign (pilcrow)
  // 0x00, 0x66, 0x89, 0x95, 0x6A, // 0x15 paragraph sign original
  0x00, 0x26, 0x49, 0x55, 0x28, // 0x15 paragraph sign
  0x60, 0x60, 0x60, 0x60, 0x60, // 0x16 double underline
  0x54, 0x62, 0x7F, 0x62, 0x54, // 0x17 up/down arrow with underline
  0x08, 0x04, 0x7E, 0x04, 0x08, // 0x18 up arrow
  0x10, 0x20, 0x7E, 0x20, 0x10, // 0x19 down arrow
  0x08, 0x08, 0x2A, 0x1C, 0x08, // 0x1A right arrow
  0x08, 0x1C, 0x2A, 0x08, 0x08, // 0x1B left arrow
  0x14, 0x3E, 0x55, 0x41, 0x22, // 0x1C euro
  0x63, 0x75, 0x69, 0x75, 0x63, // 0x1D timer
  0x30, 0x38, 0x3E, 0x38, 0x30, // 0x1E up triangle
  0x06, 0x0E, 0x3E, 0x0E, 0x06, // 0x1F down triangle
  0x00, 0x00, 0x00, 0x00, 0x00, // 0x20 space
  0x00, 0x00, 0x5F, 0x00, 0x00, // 0x21 !
  0x00, 0x07, 0x00, 0x07, 0x00, // 0x22 "
  0x14, 0x7F, 0x14, 0x7F, 0x14, // 0x23 #
  0x24, 0x2A, 0x7F, 0x2A, 0x12, // 0x24 $
  0x23, 0x13, 0x08, 0x64, 0x62, // 0x25 %
  0x36, 0x49, 0x56, 0x20, 0x50, // 0x26 &
  0x00, 0x08, 0x07, 0x03, 0x00, // 0x27 ´
  0x00, 0x1C, 0x22, 0x41, 0x00, // 0x28 (
  0x00, 0x41, 0x22, 0x1C, 0x00, // 0x29 )
  0x2A, 0x1C, 0x7F, 0x1C, 0x2A, // 0x2A *
  0x08, 0x08, 0x3E, 0x08, 0x08, // 0x2B +
  0x00, 0x80, 0x70, 0x30, 0x00, // 0x2C ,
  0x08, 0x08, 0x08, 0x08, 0x08, // 0x2D -
  0x00, 0x00, 0x60, 0x60, 0x00, // 0x2E .
  0x20, 0x10, 0x08, 0x04, 0x02, // 0x2F /
  0x3E, 0x51, 0x49, 0x45, 0x3E, // 0x30 0
  0x00, 0x42, 0x7F, 0x40, 0x00, // 0x31 1
  0x72, 0x49, 0x49, 0x49, 0x46, // 0x32 2
  0x21, 0x41, 0x49, 0x4D, 0x33, // 0x33 3
  0x18, 0x14, 0x12, 0x7F, 0x10, // 0x34 4
  0x27, 0x45, 0x45, 0x45, 0x39, // 0x35 5
  0x3C, 0x4A, 0x49, 0x49, 0x31, // 0x36 6
  0x41, 0x21, 0x11, 0x09, 0x07, // 0x37 7
  0x36, 0x49, 0x49, 0x49, 0x36, // 0x38 8
  0x46, 0x49, 0x49, 0x29, 0x1E, // 0x39 9
  0x00, 0x00, 0x14, 0x00, 0x00, // 0x3A :
  0x00, 0x40, 0x34, 0x00, 0x00, // 0x3B ;
  0x00, 0x08, 0x14, 0x22, 0x41, // 0x3C <
  0x14, 0x14, 0x14, 0x14, 0x14, // 0x3D =
  0x00, 0x41, 0x22, 0x14, 0x08, // 0x3E >
  0x02, 0x01, 0x59, 0x09, 0x06, // 0x3F ?
  0x3E, 0x41, 0x5D, 0x59, 0x4E, // 0x40 @
  0x7C, 0x12, 0x11, 0x12, 0x7C, // 0x41 A
  0x7F, 0x49, 0x49, 0x49, 0x36, // 0x42 B 
  0x3E, 0x41, 0x41, 0x41, 0x22, // 0x43 C
  0x7F, 0x41, 0x41, 0x41, 0x3E, // 0x44 D
  0x7F, 0x49, 0x49, 0x49, 0x41, // 0x45 E
  0x7F, 0x09, 0x09, 0x09, 0x01, // 0x46 F
  0x3E, 0x41, 0x41, 0x51, 0x73, // 0x47 G
  0x7F, 0x08, 0x08, 0x08, 0x7F, // 0x48 H
  0x00, 0x41, 0x7F, 0x41, 0x00, // 0x49 I
  0x20, 0x40, 0x41, 0x3F, 0x01, // 0x4A J
  0x7F, 0x08, 0x14, 0x22, 0x41, // 0x4B K
  0x7F, 0x40, 0x40, 0x40, 0x40, // 0x4C L
  0x7F, 0x02, 0x1C, 0x02, 0x7F, // 0x4D M
  0x7F, 0x04, 0x08, 0x10, 0x7F, // 0x4E N
  0x3E, 0x41, 0x41, 0x41, 0x3E, // 0x4F O
  0x7F, 0x09, 0x09, 0x09, 0x06, // 0x50 P
  0x3E, 0x41, 0x51, 0x21, 0x5E, // 0x51 Q
  0x7F, 0x09, 0x19, 0x29, 0x46, // 0x52 R
  0x26, 0x49, 0x49, 0x49, 0x32, // 0x53 S
  0x03, 0x01, 0x7F, 0x01, 0x03, // 0x54 T
  0x3F, 0x40, 0x40, 0x40, 0x3F, // 0x55 U
  0x1F, 0x20, 0x40, 0x20, 0x1F, // 0x56 V
  0x3F, 0x40, 0x38, 0x40, 0x3F, // 0x57 W
  0x63, 0x14, 0x08, 0x14, 0x63, // 0x58 X
  0x03, 0x04, 0x78, 0x04, 0x03, // 0x59 Y
  0x61, 0x59, 0x49, 0x4D, 0x43, // 0x5A Z
  0x00, 0x7F, 0x41, 0x41, 0x41, // 0x5B [
  0x02, 0x04, 0x08, 0x10, 0x20, // 0x5C '\'
  0x00, 0x41, 0x41, 0x41, 0x7F, // 0x5D ]
  0x04, 0x02, 0x01, 0x02, 0x04, // 0x5E ^
  0x40, 0x40, 0x40, 0x40, 0x40, // 0x5F _
  0x00, 0x03, 0x07, 0x08, 0x00, // 0x60 `
  0x20, 0x54, 0x54, 0x78, 0x40, // 0x61 a
  0x7F, 0x28, 0x44, 0x44, 0x38, // 0x62 b
  0x38, 0x44, 0x44, 0x44, 0x00, // 0x63 c
  0x38, 0x44, 0x44, 0x28, 0x7F, // 0x64 d
  0x38, 0x54, 0x54, 0x54, 0x18, // 0x65 e
  0x00, 0x08, 0x7E, 0x09, 0x02, // 0x66 f
  // 0x18, 0xA4, 0xA4, 0x9C, 0x78, // 0x67 g (old)
  0x0C, 0x52, 0x52, 0x4E, 0x3C, // 0x67 g 
  0x7F, 0x08, 0x04, 0x04, 0x78, // 0x68 h
  0x00, 0x44, 0x7D, 0x40, 0x00, // 0x69 i
  0x20, 0x40, 0x40, 0x3D, 0x00, // 0x6A j
  0x7F, 0x10, 0x28, 0x44, 0x00, // 0x6B k
  0x00, 0x41, 0x7F, 0x40, 0x00, // 0x6C l
  0x7C, 0x04, 0x78, 0x04, 0x78, // 0x6D m
  0x7C, 0x08, 0x04, 0x04, 0x78, // 0x6E n
  0x38, 0x44, 0x44, 0x44, 0x38, // 0x6F o
  0xFC, 0x18, 0x24, 0x24, 0x18, // 0x70 p
  // 0x18, 0x24, 0x24, 0x18, 0xFC, // 0x71 q (old)
  0x0C, 0x12, 0x12, 0x0C, 0x7E, // 0x71 q
  0x7C, 0x08, 0x04, 0x04, 0x08, // 0x72 r
  0x48, 0x54, 0x54, 0x54, 0x24, // 0x73 s
  0x04, 0x04, 0x3F, 0x44, 0x24, // 0x74 t
  0x3C, 0x40, 0x40, 0x20, 0x7C, // 0x75 u
  0x1C, 0x20, 0x40, 0x20, 0x1C, // 0x76 v
  0x3C, 0x40, 0x30, 0x40, 0x3C, // 0x77 w
  0x44, 0x28, 0x10, 0x28, 0x44, // 0x78 x
  0x0C, 0x10, 0x10, 0x50, 0x3C, // 0x79 y
  0x44, 0x64, 0x54, 0x4C, 0x44, // 0x7A z
  0x00, 0x08, 0x36, 0x41, 0x00, // 0x7B {
  0x00, 0x00, 0x7F, 0x00, 0x00, // 0x7C |
  0x00, 0x41, 0x36, 0x08, 0x00, // 0x7D }
  0x02, 0x01, 0x02, 0x04, 0x02, // 0x7E ~
  0x3C, 0x26, 0x23, 0x26, 0x3C, // 0x7F house
  0x1E, 0xA1, 0xA1, 0x61, 0x12, // 0x81
  0x3A, 0x40, 0x40, 0x20, 0x7A, // 0x81 ü
  0x38, 0x54, 0x54, 0x55, 0x59, // 0x82
  0x21, 0x55, 0x55, 0x79, 0x41, // 0x83
  0x21, 0x54, 0x54, 0x78, 0x41, // 0x84 ä
  0x21, 0x55, 0x54, 0x78, 0x40, // 0x85
  0x20, 0x54, 0x55, 0x79, 0x40, // 0x86
  0x0C, 0x1E, 0x52, 0x72, 0x12, // 0x87
  0x39, 0x55, 0x55, 0x55, 0x59, // 0x88
  0x39, 0x54, 0x54, 0x54, 0x59, // 0x89
  0x39, 0x55, 0x54, 0x54, 0x58, // 0x8A
  0x00, 0x00, 0x45, 0x7C, 0x41, // 0x8B
  0x00, 0x02, 0x45, 0x7D, 0x42, // 0x8C
  0x00, 0x01, 0x45, 0x7C, 0x40, // 0x8D
  0xF0, 0x29, 0x24, 0x29, 0xF0, // 0x8E Ä
  0xF0, 0x28, 0x25, 0x28, 0xF0, // 0x8F
  0x7C, 0x54, 0x55, 0x45, 0x00, // 0x90
  0x20, 0x54, 0x54, 0x7C, 0x54, // 0x91
  0x7C, 0x0A, 0x09, 0x7F, 0x49, // 0x92
  0x32, 0x49, 0x49, 0x49, 0x32, // 0x93
  0x32, 0x48, 0x48, 0x48, 0x32, // 0x94 ö
  0x32, 0x4A, 0x48, 0x48, 0x30, // 0x95
  0x3A, 0x41, 0x41, 0x21, 0x7A, // 0x96
  0x3A, 0x42, 0x40, 0x20, 0x78, // 0x97
  0x00, 0x9D, 0xA0, 0xA0, 0x7D, // 0x98
  //0x39, 0x44, 0x44, 0x44, 0x39, // 0x99 Ö (old)
  0x3D, 0x42, 0x42, 0x42, 0x3D, // 0x99 Ö
  0x3D, 0x40, 0x40, 0x40, 0x3D, // 0x9A Ü
  0x3C, 0x24, 0xFF, 0x24, 0x24, // 0x9B cent
  0x48, 0x7E, 0x49, 0x43, 0x66, // 0x9C pound
  0x2B, 0x2F, 0xFC, 0x2F, 0x2B, // 0x9D
  0xFF, 0x09, 0x29, 0xF6, 0x20, // 0x9E
  0xC0, 0x88, 0x7E, 0x09, 0x03, // 0x9F
  0x20, 0x54, 0x54, 0x79, 0x41, // 0xA0
  0x00, 0x00, 0x44, 0x7D, 0x41, // 0xA1
  0x30, 0x48, 0x48, 0x4A, 0x32, // 0xA2
  0x38, 0x40, 0x40, 0x22, 0x7A, // 0xA3
  0x00, 0x7A, 0x0A, 0x0A, 0x72, // 0xA4
  0x7D, 0x0D, 0x19, 0x31, 0x7D, // 0xA5
  0x26, 0x29, 0x29, 0x2F, 0x28, // 0xA6
  0x26, 0x29, 0x29, 0x29, 0x26, // 0xA7
  0x30, 0x48, 0x4D, 0x40, 0x20, // 0xA8
  0x38, 0x08, 0x08, 0x08, 0x08, // 0xA9
  0x08, 0x08, 0x08, 0x08, 0x38, // 0xAA
  0x2F, 0x10, 0xC8, 0xAC, 0xBA, // 0xAB
  0x2F, 0x10, 0x28, 0x34, 0xFA, // 0xAC
  0x00, 0x00, 0x7B, 0x00, 0x00, // 0xAD exclamation mark upside down
  0x08, 0x14, 0x2A, 0x14, 0x22, // 0xAE less less
  0x22, 0x14, 0x2A, 0x14, 0x08, // 0xAF greater greater
  0xAA, 0x00, 0x55, 0x00, 0xAA, // 0xB0  
  0xAA, 0x55, 0xAA, 0x55, 0xAA, // 0xB1
  0xAA, 0x55, 0xAA, 0x55, 0xAA, // 0xB2
  0x00, 0x00, 0x00, 0xFF, 0x00, // 0xB3
  0x10, 0x10, 0x10, 0xFF, 0x00, // 0xB4
  0x14, 0x14, 0x14, 0xFF, 0x00, // 0xB5
  0x10, 0x10, 0xFF, 0x00, 0xFF, // 0xB6
  0x10, 0x10, 0xF0, 0x10, 0xF0, // 0xB7
  0x14, 0x14, 0x14, 0xFC, 0x00, // 0xB8
  0x14, 0x14, 0xF7, 0x00, 0xFF, // 0xB9
  0x00, 0x00, 0xFF, 0x00, 0xFF, // 0xBA
  0x14, 0x14, 0xF4, 0x04, 0xFC, // 0xBB
  0x14, 0x14, 0x17, 0x10, 0x1F, // 0xBC
  0x10, 0x10, 0x1F, 0x10, 0x1F, // 0xBD
  0x14, 0x14, 0x14, 0x1F, 0x00, // 0xBE
  0x10, 0x10, 0x10, 0xF0, 0x00, // 0xBF
  0x00, 0x00, 0x00, 0x1F, 0x10, // 0xC0
  0x10, 0x10, 0x10, 0x1F, 0x10, // 0xC1
  0x10, 0x10, 0x10, 0xF0, 0x10, // 0xC2
  0x00, 0x00, 0x00, 0xFF, 0x10, // 0xC3
  0x10, 0x10, 0x10, 0x10, 0x10, // 0xC4
  0x10, 0x10, 0x10, 0xFF, 0x10, // 0xC5
  0x00, 0x00, 0x00, 0xFF, 0x14, // 0xC6
  0x00, 0x00, 0xFF, 0x00, 0xFF, // 0xC7
  0x00, 0x00, 0x1F, 0x10, 0x17, // 0xC8
  0x00, 0x00, 0xFC, 0x04, 0xF4, // 0xC9
  0x14, 0x14, 0x17, 0x10, 0x17, // 0xCA
  0x14, 0x14, 0xF4, 0x04, 0xF4, // 0xCB
  0x00, 0x00, 0xFF, 0x00, 0xF7, // 0xCC
  0x14, 0x14, 0x14, 0x14, 0x14, // 0xCD
  0x14, 0x14, 0xF7, 0x00, 0xF7, // 0xCE
  0x14, 0x14, 0x14, 0x17, 0x14, // 0xCF
  0x10, 0x10, 0x1F, 0x10, 0x1F, // 0xD0
  0x14, 0x14, 0x14, 0xF4, 0x14, // 0xD1
  0x10, 0x10, 0xF0, 0x10, 0xF0, // 0xD2
  0x00, 0x00, 0x1F, 0x10, 0x1F, // 0xD3
  0x00, 0x00, 0x00, 0x1F, 0x14, // 0xD4
  0x00, 0x00, 0x00, 0xFC, 0x14, // 0xD5
  0x00, 0x00, 0xF0, 0x10, 0xF0, // 0xD6
  0x10, 0x10, 0xFF, 0x10, 0xFF, // 0xD7
  0x14, 0x14, 0x14, 0xFF, 0x14, // 0xD8
  0x10, 0x10, 0x10, 0x1F, 0x00, // 0xD9
  0x00, 0x00, 0x00, 0xF0, 0x10, // 0xDA
  0xFF, 0xFF, 0xFF, 0xFF, 0xFF, // 0xDB
  0xF0, 0xF0, 0xF0, 0xF0, 0xF0, // 0xDC
  0xFF, 0xFF, 0xFF, 0x00, 0x00, // 0xDD
  0x00, 0x00, 0x00, 0xFF, 0xFF, // 0xDE
  0x0F, 0x0F, 0x0F, 0x0F, 0x0F, // 0xDF
  0x38, 0x44, 0x44, 0x38, 0x44, // 0xE0 alpha
  0x7C, 0x2A, 0x2A, 0x3E, 0x14, // 0xE1 beta / ß
  0x7E, 0x02, 0x02, 0x06, 0x06, // 0xE2 Gamma
  0x02, 0x7E, 0x02, 0x7E, 0x02, // 0xE3 Pi
  0x63, 0x55, 0x49, 0x41, 0x63, // 0xE4 Sigma
  0x38, 0x44, 0x44, 0x3C, 0x04, // 0xE5 sigma
  0x40, 0x7E, 0x20, 0x1E, 0x20, // 0xE6 mu
  0x06, 0x02, 0x7E, 0x02, 0x02, // 0xE7 tau
  0x99, 0xA5, 0xE7, 0xA5, 0x99, // 0xE8 Phi
  0x1C, 0x2A, 0x49, 0x2A, 0x1C, // 0xE9 theta
  0x4C, 0x72, 0x01, 0x72, 0x4C, // 0xEA Omega
  0x30, 0x4A, 0x4D, 0x4D, 0x30, // 0xEB delta
  0x30, 0x48, 0x78, 0x48, 0x30, // 0xEC phi?
  0xBC, 0x62, 0x5A, 0x46, 0x3D, // 0xED empty set
  0x3E, 0x49, 0x49, 0x49, 0x00, // 0xEE epsilon
  0x7E, 0x01, 0x01, 0x01, 0x7E, // 0xEF
  0x2A, 0x2A, 0x2A, 0x2A, 0x2A, // 0xF0 eqivalence
  0x44, 0x44, 0x5F, 0x44, 0x44, // 0xF1 plus minus
  0x40, 0x51, 0x4A, 0x44, 0x40, // 0xF2 greater or equal
  0x40, 0x44, 0x4A, 0x51, 0x40, // 0xF3 less or equal
  0x00, 0x00, 0xFF, 0x01, 0x03, // 0xF4
  0xE0, 0x80, 0xFF, 0x00, 0x00, // 0xF5
  0x08, 0x08, 0x6B, 0x6B, 0x08, // 0xF6 divided by
  0x36, 0x12, 0x36, 0x24, 0x36, // 0xF7 approx
  0x06, 0x0F, 0x09, 0x0F, 0x06, // 0xF8 degree
  0x00, 0x00, 0x18, 0x18, 0x00, // 0xF9 big dot
  0x00, 0x00, 0x10, 0x10, 0x00, // 0xFA small dot
  0x30, 0x40, 0xFF, 0x01, 0x01, // 0xFB square root
  0x00, 0x1F, 0x01, 0x01, 0x1E, // 0xFC
  0x00, 0x19, 0x1D, 0x17, 0x12, // 0xFD power of two
  0x00, 0x3C, 0x3C, 0x3C, 0x3C, // 0xFE blob
  0x00, 0x00, 0x00, 0x00, 0x00, // 0xFF
};
#endif
